File: js/modules/blockchain.js
window.API = window.API || 'http://localhost:8000/api';

let blocks = [];

// Load from Python on start
async function loadChain() {
    try {
        const res = await fetch(`${API}/blockchain`);
        const data = await res.json();
        
        console.log("Python response:", data); // DEBUG
        
        blocks = data.blocks || []; // Ensure it's an array
        
        if (blocks.length === 0) {
            logSys('t-warn', 'No blocks returned from Python');
            return;
        }
        
        renderChain();
        logSys('t-ok', `Loaded ${blocks.length} blocks from Python`);
        
    } catch (err) {
        console.error("Load error:", err);
        logSys('t-err', 'Backend offline - start Python: python backend/app.py');
    }
}

// Render with null checks
window.renderChainOriginal = renderChain;
function renderChain() {
    const chainViz = document.getElementById('chainViz');
    const chainTbody = document.getElementById('chainTbody');
    const chainAlert = document.getElementById('chainAlert');
    const blockCount = document.getElementById('blockCount');
    
    if (!chainViz || !chainTbody || !chainAlert) {
        console.error("Missing DOM elements");
        return;
    }
    
    if (!blocks || blocks.length === 0) {
        chainAlert.innerHTML = '<span id="chainAlertIcon">⚠</span> <span id="chainAlertText">No blocks to display</span>';
        chainAlert.className = 'alert alert-orange';
        if (blockCount) blockCount.textContent = '0 blocks';
        return;
    }

    // Update block count
    if (blockCount) blockCount.textContent = `${blocks.length} blocks`;

    // Visualization - Horizontal chain
    chainViz.innerHTML = blocks.map((b, i) => `
        ${i > 0 ? '<div class="chain-arrow">→</div>' : ''}
        <div class="block-item">
            <div class="block-num">BLOCK #${b.num}</div>
            <div class="block-hash">${b.hash.substring(0, 16)}...</div>
            <div class="block-data">${b.data}</div>
            <div class="block-nonce">nonce: ${b.nonce}</div>
        </div>
    `).join('');

    // Table
    chainTbody.innerHTML = blocks.map(b => `
        <tr>
            <td class="mono" style="color:var(--accent)">${b.num}</td>
            <td class="mono" style="font-size:10px">${b.hash}</td>
            <td class="mono" style="font-size:10px;color:var(--text3)">${b.prev}</td>
            <td style="font-size:12px;max-width:160px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap">${b.data}</td>
            <td class="mono">${b.nonce}</td>
            <td class="mono" style="color:var(--text2);font-size:11px">${b.ts}</td>
        </tr>
    `).join('');

    chainAlert.innerHTML = '<span id="chainAlertIcon">✓</span> <span id="chainAlertText">Chain integrity verified — all ' + blocks.length + ' blocks valid</span>';
    chainAlert.className = 'alert alert-green';
}

// Mine block
async function mineBlock() {
    const inputEl = document.getElementById('mineInput');
    const iocData = inputEl ? inputEl.value.trim() : '';

    if (!iocData) {
        logSys('t-warn', 'Enter an IOC to mine into the block');
        return;
    }

    logSys('t-ok', 'Mining... (Python calculating proof-of-work)');

    try {
        const res = await fetch(`${window.API}/blockchain/mine`, {
            method: 'POST',
            headers: { 
                'Content-Type': 'application/json',
                'X-API-Key': 'sentra-dev-key'
            },
            body: JSON.stringify({ data: `IOC: ${iocData}` })
        });

        const data = await res.json();
        console.log("Mine response:", data); // DEBUG

        if (data.error) {
            logSys('t-err', data.error);
            return;
        }

        if (data.block) {
            blocks.push(data.block);
            renderChain();
            inputEl.value = ''; // clear after mining
            logSys('t-ok', `Block #${data.block.num} mined (nonce: ${data.block.nonce}, attempts: ${data.attempts})`);
        }
    } catch (err) {
        console.error("Mine error:", err);
        logSys('t-err', 'Mining failed: ' + err.message);
    }
}

// Initialize on page load
document.addEventListener('DOMContentLoaded', function() {
    // Only load chain if we're on the blockchain page
    if (document.getElementById('page-chain')) {
        loadChain();
    }
});

File: backend/blockchain.py


from fastapi import APIRouter
import hashlib
import time

router = APIRouter()

# ── In-memory chain (replace with DB later) ──────────
blockchain = [{
    "num":   0,
    "hash":  "0000a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8",
    "prev":  "0" * 64,
    "data":  "GENESIS BLOCK",
    "nonce": 0,
    "ts":    "2024-01-01 00:00:00"
}]


def sha256_hash(num, prev, data, nonce, ts) -> str:
    text = f"{num}{prev}{data}{nonce}{ts}"
    return hashlib.sha256(text.encode()).hexdigest()


# ── Endpoints ─────────────────────────────────────────

@router.get("")
def get_chain():
    """Return full blockchain."""
    return {"blocks": blockchain, "length": len(blockchain)}


@router.post("/mine")
def mine_block(data: dict):
    """
    Mine a new block with proof-of-work (hash must start with '0000').
    Body: { "data": "some string" }
    """
    last  = blockchain[-1]
    ts    = time.strftime("%Y-%m-%d %H:%M:%S")
    nonce = 0

    while True:
        h = sha256_hash(last["num"] + 1, last["hash"], data.get("data", "BLOCK"), nonce, ts)
        if h.startswith("0000"):
            break
        nonce += 1
        if nonce > 200000:
            return {"error": "Mining timeout — nonce exceeded 200k"}

    new_block = {
        "num":   last["num"] + 1,
        "hash":  h,
        "prev":  last["hash"],
        "data":  data.get("data", "BLOCK"),
        "nonce": nonce,
        "ts":    ts
    }
    blockchain.append(new_block)
    return {"block": new_block, "mined": True, "attempts": nonce}